src/main/perl/RuleBasedEditor.pm

```diff
@@ -0,0 +1,1215 @@
+        "GLOBUS_THREAD_MODEL" => "globusThreadModel:dpm;".LINE_FORMAT_ENV_VAR,
+        "DISKFLAGS" =>"DiskFlags:dpm;".LINE_FORMAT_SH_VAR.";".LINE_VALUE_ARRAY,
+       );
+
+For more comprehensive examples of rules, look at L<ncm-dpmlfc> or L<ncm-xrootd> source code in
+configuration-modules-grid repository.
+
+=cut
+
+
+package CAF::RuleBasedEditor;
+
+use strict;
+use warnings;
+use vars qw($EC);
+$EC = LC::Exception::Context->new->will_store_all;
+
+use parent qw(CAF::FileEditor Exporter);
+
+use EDG::WP4::CCM::Element;
+
+use Readonly;
+
+use Encode qw(encode_utf8);
+
+# Constants from FileEditor
+use CAF::FileEditor qw(BEGINNING_OF_FILE ENDING_OF_FILE);
+
+
+=pod
+
+=head2 Rule Constants
+
+The constants described here are used to build the rules. All these
+constants are exported. Add the following to use them:
+
+    use RuleBasedEditor qw(:rule_constants);
+
+There is a different group of constants for each part of the rule.
+
+
+=head3  LINE_FORMAT_xxx: general syntax of the line
+
+=over
+
+=item *
+
+LINE_FORMAT_SH_VAR:         key=val (e.g. SH shell family). A comment is added at the 
+end of the line if it is modified by L<CAF::RuleBasedEditor>.
+
+=item *
+
+LINE_FORMAT_ENV_VAR:        export key=val (e.g. SH shell family). A comment is added at the 
+end of the line if it is modified by L<CAF::RuleBasedEditor>.
+
+
+=item *
+
+LINE_FORMAT_KEY_VAL:        key val (e.g. Xrootd, Apache)
+
+=item *
+LINE_FORMAT_KEY_VAL_SETENV: setenv key=val  (used by Xrootd in particular)
+
+=item *
+
+LINE_FORMAT_KEY_VAL_SET:    set key=val  (CSH shell variable)
+
+=back
+
+=cut
+
+use enum qw(
+    LINE_FORMAT_SH_VAR=1
+    LINE_FORMAT_ENV_VAR
+    LINE_FORMAT_KEY_VAL
+    LINE_FORMAT_KEY_VAL_SETENV
+    LINE_FORMAT_KEY_VAL_SET
+    );
+
+=pod
+
+=head3 
+
+LINE_VALUE_xxx: how to interpret the configuration value
+
+=over
+
+=item
+
+LINE_VALUE_AS_IS: take the value as it is, do not attempt any conversion
+
+=item
+
+LINE_VALUE_BOOLEAN: interpret the value as a boolean rendered as C<yes> or C<no>
+
+=item
+
+LINE_VALUE_ARRAY: the value is an array. Rendering controlled by LINE_VALUE_OPT_xxx constants.
+
+=item
+
+LINE_VALUE_HASH_KEYS: the value is hash whose keys are the value. Rendering similar to arrays.
+
+=item
+
+LINE_VALUE_STRING_HASH: the value is a hash of string. Rendering controlled by LINE_VALUE_OPT_xxx constants.
+
+=item
+
+LINE_VALUE_INSTANCE_PARAMS: specific to L<ncm-xrootd>
+
+=back
+
+=cut
+
+use enum qw(
+    LINE_VALUE_AS_IS
+    LINE_VALUE_BOOLEAN
+    LINE_VALUE_ARRAY
+    LINE_VALUE_HASH_KEYS
+    LINE_VALUE_STRING_HASH
+    LINE_VALUE_INSTANCE_PARAMS
+    );
+
+=pod
+
+=head3 LINE_VALUE_OPT_xxx: options for rendering the value
+
+These options mainly apply to lists and hashes and are interpreted as a bitmask.
+
+=over
+
+=item
+
+LINE_VALUE_OPT_SINGLE: each value must be a separate instance of the keyword (multiple lines)
+
+=item
+
+LINE_VALUE_OPT_UNIQUE: each values are concataneted as a space-separated string
```

Comment:
typo `concataneted` -> `concatenated` (and `all values` instead of `each values`)